test/common/secret/secret_manager_impl_test.cc

```diff
@@ -69,6 +74,51 @@ name: "abc.com"
                             "Secret type not implemented");
 }
 
+TEST_F(SecretManagerImplTest, SdsDynamicSecretUpdateSuccess) {
+  Server::MockInstance server;
+  std::unique_ptr<SecretManager> secret_manager(new SecretManagerImpl());
+
+  NiceMock<Server::Configuration::MockTransportSocketFactoryContext> secret_context;
+
+  envoy::api::v2::core::ConfigSource config_source;
+  {
+    NiceMock<LocalInfo::MockLocalInfo> local_info;
+    NiceMock<Event::MockDispatcher> dispatcher;
+    NiceMock<Runtime::MockRandomGenerator> random;
+    Stats::IsolatedStoreImpl stats;
+    NiceMock<Upstream::MockClusterManager> cluster_manager;
+    NiceMock<Init::MockManager> init_manager;
+    EXPECT_CALL(secret_context, localInfo()).WillOnce(ReturnRef(local_info));
+    EXPECT_CALL(secret_context, dispatcher()).WillOnce(ReturnRef(dispatcher));
+    EXPECT_CALL(secret_context, random()).WillOnce(ReturnRef(random));
+    EXPECT_CALL(secret_context, stats()).WillOnce(ReturnRef(stats));
+    EXPECT_CALL(secret_context, clusterManager()).WillOnce(ReturnRef(cluster_manager));
+    EXPECT_CALL(secret_context, initManager()).WillRepeatedly(Return(&init_manager));
+
+    auto secret_provider = secret_manager->findOrCreateTlsCertificateProvider(
+        config_source, "abc.com", secret_context);
+    std::string yaml =
+        R"EOF(
+name: "abc.com"
+tls_certificate:
+  certificate_chain:
+    filename: "{{ test_rundir }}/test/common/ssl/test_data/selfsigned_cert.pem"
+  private_key:
+    filename: "{{ test_rundir }}/test/common/ssl/test_data/selfsigned_key.pem"
+  )EOF";
+    Protobuf::RepeatedPtrField<envoy::api::v2::auth::Secret> secret_resources;
+    auto secret_config = secret_resources.Add();
+    MessageUtil::loadFromYaml(TestEnvironment::substitute(yaml), *secret_config);
+    std::dynamic_pointer_cast<SdsApi>(secret_provider)->onConfigUpdate(secret_resources, "");
```

Comment:
Is there any way to refactor this so we don't have to do this cast?

Reply:
I don't find a way to refactor this. The onConfigUpdate() is inherited from SubscriptionCallbacks, not SecretProvider.

Reply:
Yeah I think there is no way to refactor this without dynamic_cast, but this could be `dynamic_cast<SdsApi&>(*secret_provider).onConfigUpdate`